File: C#/Logic-Implementation/Evolve.TollCalculator/TollCalculator.cs
using Evolve.TollCalculator.Extensions;
using Evolve.TollCalculator.Models;
using System;
using System.Linq;

namespace Evolve.TollCalculator
{
    public class TollCalculator
    {
        /**
         * Calculate the total toll fee for one day
         *
         * @param vehicle - the vehicle
         * @param dates   - date and time of all passes on one day
         * @return - the total toll fee for that day
         */
        public int GetTollFeeByDate(Vehicle vehicle, DateTime[] dates)
        {
            int totalFee = 0;
            if (dates.Length > 0)
            {
                var datesSorted = dates.OrderBy(d => d).ToList();
                DateTime intervalStart = datesSorted[0];

                foreach (DateTime date in datesSorted)
                {
                    int nextFee = GetTollFeeByDate(date, vehicle);
                    int tempFee = GetTollFeeByDate(intervalStart, vehicle);

                    TimeSpan span = date - intervalStart;
                    int diffInMinutes = (int)span.TotalMinutes;

                    if (diffInMinutes <= 60)
                    {
                        if (totalFee > 0)
                        {
                            totalFee -= tempFee;
                        };

                        if (nextFee >= tempFee)
                        {
                            tempFee = nextFee;
                        };

                        totalFee += tempFee;
                    }
                    else
                    {
                        totalFee += nextFee;
                    }

                    intervalStart = date;
                }
            }

            if (totalFee > 60) { totalFee = 60; };
            return totalFee;
        }

        /// <summary>
        /// Get Toll fee by 
        /// </summary>
        /// <param name="date"></param>
        /// <param name="vehicle"></param>
        /// <returns></returns>
        public int GetTollFeeByDate(DateTime date, Vehicle vehicle)
        {
            if (HolidayExtension.IsTollFreeDate(date) || VehicleExtension.IsTollFreeVehicle(vehicle)) 
            { 
                return 0; 
            }

            if (date.IsBetween(new TimeSpan(6, 0, 0), new TimeSpan(6, 29, 59))
            || date.IsBetween(new TimeSpan(8, 30, 0), new TimeSpan(14, 59, 0))
            || date.IsBetween(new TimeSpan(18, 0, 0), new TimeSpan(18, 29, 0)))
            {
                return 8;
            }
            else if (date.IsBetween(new TimeSpan(6, 30, 0), new TimeSpan(6, 59, 0))
                || date.IsBetween(new TimeSpan(8, 0, 0), new TimeSpan(8, 29, 0))
                || date.IsBetween(new TimeSpan(15, 0, 0), new TimeSpan(15, 29, 0))
                || date.IsBetween(new TimeSpan(17, 0, 0), new TimeSpan(17, 59, 0)))
            {
                return 13;
            }
            else if (date.IsBetween(new TimeSpan(7, 0, 0), new TimeSpan(7, 59, 0))
                || date.IsBetween(new TimeSpan(15, 30, 0), new TimeSpan(16, 59, 0)))
            {
                return 18;
            }
            else
            {
                return 0;
            }
        }
    }
}

File: Java/src/main/java/org/example/TollCalculator.java
package org.example;

import org.example.config.TimeslotFees;
import org.example.config.TollConfiguration;
import org.example.data.Vehicle;
import org.example.exception.APIBadRequestException;
import org.example.exception.ParameterNotFoundException;

import java.text.MessageFormat;
import java.time.DayOfWeek;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.LocalTime;
import java.time.MonthDay;
import java.time.temporal.ChronoField;
import java.time.temporal.ChronoUnit;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.stream.Collectors;

import lombok.extern.slf4j.Slf4j;
import org.example.util.VehicleValidation;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;


@Slf4j
@Component
public class TollCalculator {

    @Autowired
    private final TollConfiguration tollConfiguration;

    public TollCalculator(TollConfiguration tollConfiguration) {
        this.tollConfiguration = tollConfiguration;
        System.out.println(tollConfiguration);
    }

    /**
     * Calculate the total toll fee for one day
     *
     * @param vehicle    - the vehicle
     * @param datesInput - date and time of all passes on one day
     * @return - the total toll fee for that day
     */
    public int getTollFee(Vehicle vehicle, List<LocalDateTime> datesInput) {

        if (vehicle == null) {
            throw new ParameterNotFoundException("vehicle parameter is null.");
        }

        boolean isValidVehicle = VehicleValidation.isValidVehicleType(vehicle.getType(), tollConfiguration.getVehicleTypes());
        if(Boolean.FALSE.equals(isValidVehicle)){
            String formattedMessage = MessageFormat.format("Vehicle should be a valid. It should be of one of these vehicles: {0}" , tollConfiguration.getVehicleTypes());
            throw new APIBadRequestException(formattedMessage);
        }

        if (datesInput == null) {
            throw new ParameterNotFoundException("dates parameter is null.");
        }

        List<LocalDateTime> dates = datesInput.stream().sorted().collect(Collectors.toList());

        Map<LocalDate, List<LocalDateTime>> collect = dates.stream()
                .collect(Collectors.groupingBy(LocalDateTime::toLocalDate));

        if(collect.size()>1){
            throw new APIBadRequestException("Input dates are not valid as it includes the different days.");
        }

        if (isTollFreeVehicle(vehicle)) {
            log.info("***** Vehicle[{}] is free from toll tax, so returning 0 fee.", vehicle.getType());
            return 0;
        }
        log.info("***** Calculating toll-fees for Vehicle[{}] ***** ", vehicle.getType());


        int totalFees = 0;
        for (Map.Entry<LocalDate, List<LocalDateTime>> entry : collect.entrySet()) {
            totalFees = totalFees + getTollFeesPerDay(vehicle, entry.getValue());
        }
        log.info("Vehicle[{}] Final toll fees: {}", vehicle.getType(), totalFees);
        log.info("***** End processing toll-fees for Vehicle[{}] *****", vehicle.getType());
        return totalFees;

    }

    public int getTollFeesPerDay(Vehicle vehicle, List<LocalDateTime> localDateTimes) {

        LocalDateTime firstLocalDateTime = localDateTimes.get(0);
        if (isTollFreeDate(firstLocalDateTime)) return 0;

        int tempFee = getTollFee(firstLocalDateTime, vehicle);
        LocalDateTime tempDateTime = firstLocalDateTime;
        int totalFee = tempFee;
        List<LocalDateTime> dates = localDateTimes.stream().skip(1).collect(Collectors.toList());

        for (LocalDateTime localDateTime : dates) {
            int nextFee = getTollFee(localDateTime, vehicle);
            long minutes = tempDateTime.until(localDateTime, ChronoUnit.MINUTES);
            if(isMinutesLessThanHour(minutes)) {
                if (totalFee > 0) {
                    totalFee = totalFee - tempFee;
                }
                //consider maximum fees
                tempFee =  Math.max(nextFee, tempFee);
                totalFee = totalFee + tempFee;
            } else {
                //Consider next time slot
                tempDateTime = localDateTime;
                totalFee = totalFee + nextFee;
                tempFee = nextFee;
            }

            int maximumTollFees = tollConfiguration.getMaximumTollFeesPerDay();
            if (totalFee > maximumTollFees) {
                log.info("Calculated Total Fee: {} is more than the maximum toll fees, hence returning the maximum toll fees: {}", totalFee, maximumTollFees);
                totalFee = maximumTollFees;
                break;
            }
        }

        log.info("Total toll fees: {} for date: {}", totalFee, firstLocalDateTime.toLocalDate());

        return totalFee;
    }

    private boolean isMinutesLessThanHour(long minutes) {
        return minutes <= 60;
    }

    private boolean isTollFreeVehicle(Vehicle vehicle) {
        if (vehicle == null) return false;
        String vehicleType = vehicle.getType();
        boolean tollFreeVehicle = tollConfiguration.getTollFreeVehicles().stream().anyMatch(e -> e.equals(vehicleType));
        if (tollFreeVehicle) {
            log.info("Vehicle[{}] is free from toll tax.", vehicle.getType());
        }
        return tollFreeVehicle;
    }

    public int getTollFee(final LocalDateTime localDateTime, Vehicle vehicle) {

        if (isTollFreeDate(localDateTime) || isTollFreeVehicle(vehicle)) return 0;

        LocalTime localTime = localDateTime.toLocalTime();

        Optional<TimeslotFees> any = tollConfiguration.getTimeslotFees().stream().filter(
                e -> localTime.isBefore(e.getEndTime()) && localTime.isAfter(e.getStartTime())
        ).findAny();

        int tollFees = any.map(TimeslotFees::getFees).orElse(0);
        log.info("Vehicle[{}], toll fees:{} for date:{}",vehicle.getType(), tollFees, localDateTime);
        return tollFees;
    }

    private boolean isTollFreeDate(LocalDateTime localDateTime) {
        LocalDate localDate = localDateTime.toLocalDate();
        return isWeekend(localDate) || isHoliday(localDate);
    }

    private boolean isHoliday(final LocalDate localDate) {
        boolean holidayFlag = tollConfiguration.getHolidays().contains(MonthDay.from(localDate));
        if (holidayFlag) {
            log.info("Date[{}] It is on Holiday", localDate);
        }
        return holidayFlag;
    }

    private boolean isWeekend(final LocalDate ld) {
        DayOfWeek day = DayOfWeek.of(ld.get(ChronoField.DAY_OF_WEEK));
        boolean isDateOnWeekend = day == DayOfWeek.SUNDAY || day == DayOfWeek.SATURDAY;
        if (isDateOnWeekend) {
            log.info("Date[{}] It is on Weekend", ld);
        }
        return isDateOnWeekend;
    }


}

